### PR TITLE
PatchManager: a plugin to manage user patches

### PR DESCRIPTION
**Pull request by zwim**
This PR allows to enable/disable user patches.

(Draft by now, as there are unwanted l10n and lajit-launcher commits in here. I will look into that on WE)

I really would like to have it under `Plugin management` as I feel that my(?) patches are similar to plugins.

grafik

Here you can see, which patches (level)  are installed.

grafik

And here a patch on a certain level
grafik

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9970)

### CODE REVIEW

**Comment by melyux** [comment]
I like the idea. A question about patches: I wrote like 10 before realizing it was a lot harder than modifying the code, probably less performant, and given that Lua doesn't need to be compiled after changing the code, I found it a lot easier to have some git patches instead. Have you found a pro for patches that I haven't? Besides being able to do on-device updates without applying diff patches on the computer :)

File: plugins/patchmanager.koplugin/main.lua
Comment (hius07) [inline plugins/patchmanager.koplugin/main.lua:86]: Maybe call TextEditor?

Comment (roygbyte) [reply plugins/patchmanager.koplugin/main.lua:86]: There is a typo: patches spelt as patchse

**Comment by poire-z** [comment]
> Have you found a pro for patches that I haven't? Besides being able to do on-device updates without applying diff patches on the computer :)

That depends on the user abilities and the amount of patching one wants/needs :)
I'm still patching with my own single patch from the computer, so, needing a computer to update: https://github.com/koreader/koreader/pull/9104#issuecomment-1146340326
But for less heavy users, the ability to download user patches we made for them is easier. And on Android, there's no other way to patch, as we can't access the koreader directory (unless we are root).

File: plugins/patchmanager.koplugin/main.lua
Comment (poire-z) [reply plugins/patchmanager.koplugin/main.lua:86]: I think it is really meant as "patch set" (dunno if this can/should be made a single word "patchset" in correct English :)

**Comment by poire-z** [comment]
I initially felt it would be overkill to expose this user patch feature to the UI as 99.9% of the users won't really need it and may wonder what it's about (and best for them to not wonder and know about it if they don't need it).
But as it's 2 level down, why not.
(Though it could auto not-register its menu item if there are no user patch, but it's grayed out, which could be ok.)

May be it could also show which patch did run correctly, and those that did not, with an added icon, so one can quickly go there after an update to be sure all the patches survived.

File: plugins/patchmanager.koplugin/main.lua
Comment (Frenzie) [reply plugins/patchmanager.koplugin/main.lua:86]: I think it's correct but programmer jargon. Set of patches might suffice to make it less so.

**Comment by zwim** [comment]
> I initially felt it would be overkill to expose this user patch feature to the UI as 99.9% of the users won't really need it and may wonder what it's about (and best for them to not wonder and know about it if they don't need it). But as it's 2 level down, why not. (Though it could auto not-register its menu item if there are no user patch, but it's grayed out, which could be ok.)
 
I can enable the auto-registration only if the `patches`  folder exist. (We don't ship that, so a user having it, is aware of that feature.)


> May be it could also show which patch did run correctly, and those that did not, with an added icon, so one can quickly go there after an update to be sure all the patches survived.

:+1:

**Comment by poire-z** [comment]
> I can enable the auto-registration only if the patches folder exist. (We don't ship that, so a user having it, is aware of that feature.)

There would be then the "little" issue that having enabled "Patch manager" in the list of plugins, one would see nothing about it anywhere, and wonder why ? (unless we mention that in the help text for this plugin?)

**Comment by zwim** [comment]
Sorry, wrong wording from my side. I meant the plugin loading only to be done if the patches folder is there.

I show you ib the next commit

File: plugins/patchmanager.koplugin/main.lua
Comment (poire-z) [inline plugins/patchmanager.koplugin/main.lua:65]: You could probably go with Unicode symbols instead.
https://www.fileformat.info/info/unicode/block/dingbats/list.htm
https://www.fileformat.info/info/unicode/block/miscellaneous_symbols/list.htm

(In my own manichaean patch, I use https://www.fileformat.info/info/unicode/char/2605/index.htm and https://www.fileformat.info/info/unicode/char/2716/index.htm for good and evil - but a checkmark is probably better than my star in this context.)

Comment (zwim) [reply plugins/patchmanager.koplugin/main.lua:65]: I use the menu checkmark to show if a patch is enabled. 
What about 👍 to show success and 👎 for an error?

Comment (poire-z) [reply plugins/patchmanager.koplugin/main.lua:65]: Depend on how these "hand" icons look in our fonts, but it's also not really our "theme" to show such hands, we're rather on the simple symbols (checkmark, star...) side, so try not to go too wild :) (Also, the weight/density of both hands may be to similar to be quickly distinguished - we want the "fail/error" symbol to strike :)
There are also other kind of checkmark symbols that may look different enough than the one used on the left.

**Comment by poire-z** [comment]
Can you post a screenshot to see how these icons looks with our fonts ?

File: plugins/patchmanager.koplugin/main.lua
Comment (hius07) [inline plugins/patchmanager.koplugin/main.lua:154]: What if == 1?

Comment (zwim) [reply plugins/patchmanager.koplugin/main.lua:154]: Thank's, I should be ` > 0`.

**Comment by zwim** [comment]
About text, with icons
grafik

This is a more complicated one
`20-welcomeMessage.lua` is faulty
`2-touchmenu-clock.lua` is disabled
the others work
grafik

**Comment by zwim** [comment]
(Note to self: Once this is PR is done the natural-sorting function should be moved to `util` as we use it three times now.)

**Comment by poire-z** [comment]
What's this inverted A for "not yet executed" ? Some meaningful mathematical symbol ? In which case, that's too cryptic for the normal user I am :)
Better no symbol for "not executed" script, when the left checkbox is unchecked ?

(The hollow thin checkmark for success feels a bit remote to our theme, but ok, guess you went with that for their non-striking-ness.)

**Comment by mergen3107** [comment]
That symbols means “there exists”

**Comment by zwim** [comment]
The ∀ symbol shows that the patch has not been executed (yet). Possible causes are
* it is disabled. (which is clear, as it is disabled).
* it is only executed after an update `early_once`. (A user would never see that executed symbol here).
* it will be executed later `on_exit` or `before_exit` (and no executed symbol here, too).

As this all might be confusing I would suggest we drop all icons except the warning ( ⚠ ) on failure.

**Comment by NiLuJe** [comment]
> As this all might be confusing I would suggest we drop all icons except the warning ( warning ) on failure.

That sounds more straightforward, yeah ;).

File: plugins/patchmanager.koplugin/main.lua
Comment (hius07) [inline plugins/patchmanager.koplugin/main.lua:68]: Maybe ipairs?

File: plugins/patchmanager.koplugin/main.lua
Comment (hius07) [inline plugins/patchmanager.koplugin/main.lua:38]: So you scan the same folder 10 times.

File: plugins/patchmanager.koplugin/main.lua
Comment (poire-z) [inline plugins/patchmanager.koplugin/main.lua:128]: Why "certain found" ? :)
```
Patch manager allows to enable, disable and edit user patches.
(The execution time (I'd like to find another wording for that, step? it's like rc2.d/... on Unix) can not (yet) be modified by Patch manager, and have to be done manually by renaming the patch prefix.)

For more info about user patches, see
https://github.com/koreader/koreader/wiki/User-patches
```

File: plugins/patchmanager.koplugin/main.lua
Comment (poire-z) [inline plugins/patchmanager.koplugin/main.lua:132]: KOReader is a grumpy old program, and doesn't smile !

File: plugins/patchmanager.koplugin/main.lua
Comment (poire-z) [inline plugins/patchmanager.koplugin/main.lua:140]: May be move this nearer to where it is used below ?

File: plugins/patchmanager.koplugin/main.lua
Comment (zwim) [reply plugins/patchmanager.koplugin/main.lua:128]: Or "runlevel and priority" would be more precise.

File: plugins/patchmanager.koplugin/main.lua
Comment (poire-z) [inline plugins/patchmanager.koplugin/main.lua:103]: Why two _ strings, and why this constructs with T() and %1 ?

File: plugins/patchmanager.koplugin/main.lua
Comment (poire-z) [inline plugins/patchmanager.koplugin/main.lua:111]:
```lua
    local file = io.open(file_path, "rb")
    if not file then
        return ""
    end
    local file_content = file:read("*all")
    file:close()
```

See TextEditor:readFileContent(file_path) for the normal way of reading a file content :)
(unless this is on purpose and Lua io.lines somehow accepts and clean various linux/windows/mac line endings ? - dunno if I checked that when I made TextEditor :/)

File: plugins/patchmanager.koplugin/main.lua
Comment (zwim) [reply plugins/patchmanager.koplugin/main.lua:103]: That's to ease the translators work.
We have "Patches have changed." on menu taps and "Patches might have changed." on an edit.
But in both cases have the same "Current set of patches ....".

Comment (poire-z) [reply plugins/patchmanager.koplugin/main.lua:103]: @Frenzie : thoughts ?

Comment (zwim) [reply plugins/patchmanager.koplugin/main.lua:103]: Ahm, the `%1\n`  is the wrong way round. It should be `\n%1`.

Comment (poire-z) [reply plugins/patchmanager.koplugin/main.lua:103]: (I would make it simpler and plain and readable and avoid this :) and not care about the duplication.)

Comment (Frenzie) [reply plugins/patchmanager.koplugin/main.lua]: Either duplicate or use normal concat.

_("bla") .. "\n" _("moar")

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 3 of 5 files at r2, 1 of 2 files at r7, 1 of 3 files at r9, 1 of 2 files at r10, 1 of 3 files at r11, 1 of 1 files at r12, 1 of 1 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9970)* status: all files reviewed, 16 unresolved discussions (waiting on @hius07, @poire-z, and @zwim)

___
*[`plugins/patchmanager.koplugin/main.lua` line 59 at r13](https://reviewable.io/reviews/koreader/koreader/9970#-NKy20JL1ysNcXVjfxLr:-NKy20JL1ysNcXVjfxLs:b6lyx9y) ([raw file](https://github.com/koreader/koreader/blob/67139b7012cf65afd3d97c16c02c77e38471ea9e/plugins/patchmanager.koplugin/main.lua#L59)):*
> ```Lua
> 
> function PatchManager:getSubMenu(priority)
>     if #self.patches == 0 then
> ```

`priority` can be `0`, which makes this a hash, not an array; as such, the length `#` operator is unreliable. You'll probably have to go through the getTableSize (or whatever it's actually called) function from `util` ;).

(Or, possibly, the `next` trick might be enough here; e.g., `if next(self.patches) == nil`)
___
*[`plugins/patchmanager.koplugin/main.lua` line 139 at r13](https://reviewable.io/reviews/koreader/koreader/9970#-NKy3L6S70L3JBDnYo-k:-NKy3L6S70L3JBDnYo-l:b-89a4l7) ([raw file](https://github.com/koreader/koreader/blob/67139b7012cf65afd3d97c16c02c77e38471ea9e/plugins/patchmanager.koplugin/main.lua#L139)):*
> ```Lua
>         text = _("Patch manager"),
>         enabled_func = function()
>             if #self.patches == 0 then
> ```

Ditto

**Review by zwim — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9970)* status: all files reviewed, 16 unresolved discussions (waiting on @hius07, @NiLuJe, and @poire-z)

___
*[`plugins/patchmanager.koplugin/main.lua` line 59 at r13](https://reviewable.io/reviews/koreader/koreader/9970#-NKy20JL1ysNcXVjfxLr:-NKyEgvbCplD3Hkqt_6T:b-hnsqqd) ([raw file](https://github.com/koreader/koreader/blob/67139b7012cf65afd3d97c16c02c77e38471ea9e/plugins/patchmanager.koplugin/main.lua#L59)):*
<details><summary><i>Previously, NiLuJe (NiLuJe) wrote…</i></summary><blockquote>

`priority` can be `0`, which makes this a hash, not an array; as such, the length `#` operator is unreliable. You'll probably have to go through the getTableSize (or whatever it's actually called) function from `util` ;).

(Or, possibly, the `next` trick might be enough here; e.g., `if next(self.patches) == nil`)
</blockquote></details>

The `next` trick seems to work.

**Comment by NiLuJe** [comment]
> The `next` trick seems to work.

Cool. (Just noticed that it's not really a trick as it's explicitly detailed for that specific use-case in the `next` docs ;p).

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 1 of 1 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9970)* status: all files reviewed, 14 unresolved discussions (waiting on @hius07, @poire-z, and @zwim)

**Comment by hius07** [comment]
Sorry, I'd withdraw the idea of calling the TextEditor.
TextViewer on long-pressing would be fine, to check out the content of a patch.

**Comment by zwim** [comment]
> Sorry, I'd withdraw the idea of calling the TextEditor. TextViewer on long-pressing would be fine, to check out the content of a patch.

Well I like the idea now and I would want to keep it. (It costs nothing)

**Comment by zwim** [comment]
> > The `next` trick seems to work.
> 
> Cool :). (Fun fact: just noticed that it's not really a trick as it's explicitly detailed for that specific use-case in the `next` docs ;p).

Today I woke up and realized that in this special case `#self.patches == 0` would also work: 
This is because `self.patches` is just a container. If the `patches` folder does not exist, then `self.pachtes == {}`; 
if it is there, `self.patches = { {}, {}, {}, ...}` (with 10 empty arrays).

But I would want to stay with the `next` trick, as it is not wrong and a more general solution. :)

**Comment by poire-z** [comment]
So, if the patch folders does not exist, your pluding is not loaded and no code runs.
If the patch folder exists, self.patches has 10 items.
So, it's never empty and you don't need this check ? :)

**Comment by hius07** [comment]
Does anybody really enumerate patches with multi-digit numbers?
(Restricting the name to one_digit-name will simplify a lot. (Rare) execution ordering can be solved with some digits **after** the minus).

**Comment by poire-z** [comment]
> Does anybody really enumerate patches with multi-digit numbers? (Restricting the name to one_digit-name will simplify a lot. (Rare) execution ordering can be solved with some digits **after** the minus, without "natural sorting" etc., the author will take care of it himself).

Welcome to the wonderful [Land of OWZ](https://en.wikipedia.org/wiki/Land_of_Oz) (Over-engineered World of Zwim) !

**Comment by Frenzie** [comment]
Sidenote, I've started having trouble starting [Zim](https://zim-wiki.org/) because I keep typing Zwim.

**Comment by zwim** [comment]
> So, if the patch folders does not exist, your pluding is not loaded and no code runs. If the patch folder exists, self.patches has 10 items. So, it's never empty and you don't need this check ? :)

Yes, that's right. This comes from the change (you have stimulated), that the Plugin Manager should not be visible for users not using user patches. (Which I support.)

You could state now, that I could remove the check. I would _not_ remove it, as
1st it costs nothing,
2nd who knows, how plugin manager evolves in future (better safe than sorrow),
3rd it is a plugin (where newcomers can learn the secrets** of LUA.

TODO: For the learning point of view, this special check should be documented.


**: I meant that an array is in the range [1...N] (https://www.lua.org/pil/11.1.html) but you can also use [0] which is then something different :(

**Comment by poire-z** [comment]
Well, rather than document advanced Lua tricks everywhere it is used (I think we must have done that in some places), you could just make it clearer by not using the trick and setting a `self.has_patches = true/false` and use that (which costs less and is more readable and self documenting :).
It's only when you get the table from elsewhere and you don't know if it's empty or not that you may need to use `if next(that_thingie)`

**Comment by zwim** [comment]
Ready frim my side.

Using the nil check is even mire straight forward than an additional has_patches. :))

File: frontend/ui/elements/filemanager_menu_order.lua
Comment (Frenzie) [inline frontend/ui/elements/filemanager_menu_order.lua:138]: Shouldn't that be patch_manager to be consistent with the other menu entries? ^_^

File: plugins/patchmanager.koplugin/_meta.lua
Comment (Frenzie) [inline plugins/patchmanager.koplugin/_meta.lua:5]:
```suggestion
    description = _("This plugin allows enabling, disabling or editing user patches."),
```

**Comment by Frenzie** [comment]
@zwim It can probably go into .01 unless there's a reason not to?

File: frontend/ui/elements/filemanager_menu_order.lua
Comment (zwim) [reply frontend/ui/elements/filemanager_menu_order.lua:138]: If we want that to be ultimate consistant , it should be "Patch management".
In the code and in the menu.

**Comment by zwim** [comment]
> @zwim It can probably go into .01 unless there's a reason not to?

From my side it could go into .01; it works.
But I can do your suggested changes at earliest on Saturday, and the translaturs need time, too.

File: frontend/ui/elements/filemanager_menu_order.lua
Comment (Frenzie) [reply frontend/ui/elements/filemanager_menu_order.lua:138]: Yes, it should also be management, good point. But the _ thing jumps out without even looking at the words. ;-)

**Comment by Frenzie** [comment]
Sure thing.

**Comment by zwim** [comment]
Patch manager also uses natural sorting, so I would want to wait until #10023 is merged.

**Comment by zwim** [comment]
Added #10023.

**Comment by zwim** [comment]
Ready from my side.

File: plugins/patchmanagement.koplugin/main.lua
Comment (NiLuJe) [inline plugins/patchmanagement.koplugin/main.lua:118]: leftover?

File: plugins/patchmanagement.koplugin/main.lua
Comment (NiLuJe) [inline plugins/patchmanagement.koplugin/main.lua:132]: typo: info**r**mation

Comment (zwim) [reply plugins/patchmanagement.koplugin/main.lua:132]: Thanks.

**Review by NiLuJe — APPROVED** [review]

**Comment by zwim** [comment]
Any objections to merge this?

**Review by poire-z — APPROVED** [review]
Pinging @Frenzie to have a look at the wording/translatable strings

**Review by Frenzie — APPROVED** [review]

**Comment by poire-z** [comment]
I used to rename the user patches I made out for users, but don't use myself, by adding a _ at the end:
`2-dict-auto-copy-current-definition.lua_`
Somehow, this plugin detects them, show them, allow enabling and disabling them back !
`2-dict-auto-copy-current-definition.lua.disabled`
Was this a design decision ? Or a happy side effect of overeng. ? :)

At one point, I was with both these at the same time.
```
2-dict-auto-copy-current-definition.lua
2-dict-auto-copy-current-definition.lua.disabled
```
But after disable and re-enable, there was only one copy again - and couldn't reproduce that (on that same script, I could probably on the other ones still named with a trailing `_`) - so no real issue.

**Comment by Frenzie** [comment]
> Or a happy side effect of overeng. ? :)

Sounds more like starts with "lua" instead of equals "lua". Not sure that's happy. ^_^

File: plugins/patchmanagement.koplugin/main.lua
Comment (zwim) [inline plugins/patchmanagement.koplugin/main.lua:70]: > Was this a design decision ? Or a happy side effect of overeng. ? :)

@poire-z :I think the author deliberately decided to do it hat way

Comment (zwim) [reply plugins/patchmanagement.koplugin/main.lua:70]: That depends on what you mean with well. 

`2-somestuff.lua.old` is handled as `2-somestuff.lua.disabled`.  So any file ending with `.luaxxx` is disabled.

So  `2-somestuff.lua.old` *will* show up in patch management as disabled. If you don't want to show it up you can name it as `x2-somestuff.lua` or `abc2-somestuff.lua` or `outdatedxxx2-somestuff.lua`.

If you have `2-somestuff.lua.old` and `2-somestuff.lua` in `patches` both are shown. The first as disabled, the second as enabled. This is not optimal, but reflects the current files.
If you dis/enable one of them, the behavior is not designed.

Comment (zwim) [reply plugins/patchmanagement.koplugin/main.lua:70]: So if you want to keep an old version of a patch, you can/have to rename it to `2-somestuffxxxx.lua` (I would suggest `2-somestuff-v1.lua` or `2-somestuff-20230206.lua`) and you can dis/enable it as you like.

**Pull request merged**
